File: sdks/typescript/packages/connect/src/client-auth.ts
/**
 * @author Sudomimus Contributors
 * @package Connect
 * @namespace ClientAuth
 * @description Client-auth JWT helpers for /establish
 */

import { JWTCreator } from "@sudoo/jwt";
import * as crypto from "crypto";
import {
    CLIENT_JWT_AUDIENCE,
    CLIENT_JWT_DEFAULT_LIFETIME_SECONDS,
    CLIENT_JWT_MAX_LIFETIME_SECONDS,
} from "./constants";
import type { ConnectClientAuthWithKey } from "./declare";
import { ConnectConfigError } from "./error";

export type EstablishClientJwtClaims = {

    readonly iss: string;
    readonly aud: string;
    readonly iat: number;
    readonly exp: number;
    readonly jti: string;
    readonly body_sha256: string;
};

export type BuildEstablishClientJwtClaimsOptions = {

    readonly lifetimeSeconds?: number;
    readonly jtiGenerator?: () => string;
    readonly now?: Date;
};

/**
 * Server hashes the raw HTTP body with `SHA-256` over UTF-8 bytes and
 * encodes as standard base64. Mirror that exactly — any drift means
 * `S_ClientJwtBodyHashMismatch` server-side.
 */
export const sha256Base64 = (input: string): string => {

    return crypto.createHash("sha256").update(input, "utf8").digest("base64");
};

/**
 * Build the client-auth JWT claim set without signing it. Useful for BYO
 * signers that need the claim object as input.
 */
export const buildEstablishClientJwtClaims = (
    applicationAnchor: string,
    rawBody: string,
    options: BuildEstablishClientJwtClaimsOptions = {},
): EstablishClientJwtClaims => {

    const lifetimeSeconds = options.lifetimeSeconds ?? CLIENT_JWT_DEFAULT_LIFETIME_SECONDS;

    if (lifetimeSeconds <= 0 || lifetimeSeconds > CLIENT_JWT_MAX_LIFETIME_SECONDS) {

        throw new ConnectConfigError(
            `clientAuth.lifetimeSeconds must be in (0, ${CLIENT_JWT_MAX_LIFETIME_SECONDS}]; got ${lifetimeSeconds}`,
        );
    }

    const now = options.now ?? new Date();
    const iat = Math.floor(now.getTime() / 1000);
    const exp = iat + lifetimeSeconds;
    const jti = (options.jtiGenerator ?? crypto.randomUUID)();

    return {
        iss: applicationAnchor,
        aud: CLIENT_JWT_AUDIENCE,
        iat,
        exp,
        jti,
        body_sha256: sha256Base64(rawBody),
    };
};

/**
 * Sign a client-auth JWT using a PEM-encoded RS256 private key. Returns the
 * compact JWT string; caller is responsible for prefixing it with the
 * `SudomimusClientJWT ` scheme in the `Authorization` header.
 *
 * All claims (`iss`, `aud`, `iat`, `exp`, `jti`, `body_sha256`) are emitted
 * in the JWT body (payload), not the header. This matches the server's
 * `verifyEstablishClientJwt` reader which inspects `parsed.body`. Note that
 * `@sudoo/jwt`'s `issuer`/`audience`/`issuedAt`/`expirationAt`/`identifier`
 * verbal options would route into `parsed.header` instead — so we set the
 * body fields directly and leave the verbal options unused.
 */
export const signEstablishClientJwt = (
    config: ConnectClientAuthWithKey,
    rawBody: string,
): string => {

    const claims = buildEstablishClientJwtClaims(config.applicationAnchor, rawBody, {
        lifetimeSeconds: config.lifetimeSeconds,
        jtiGenerator: config.jtiGenerator,
    });

    const creator: JWTCreator<Record<string, never>, EstablishClientJwtClaims> =
        JWTCreator.instantiate(config.privateKeyPem);

    return creator.create({
        header: {},
        body: claims,
    });
};

File: sdks/python/packages/sudomimus-connect/src/sudomimus_connect/client_auth.py
"""Client-auth JWT signing for the Connect ``/establish`` endpoint.

The server requires an ``Authorization: SudomimusClientJWT <jwt>`` header
whose claims live in the JWT *body*: ``iss`` (the application anchor),
``aud`` (``"sudomimus-connect"``), ``iat``, ``exp`` (lifetime at most 60s),
a single-use ``jti``, and ``body_sha256`` — standard base64 of
``SHA-256(rawHttpBody)`` over the exact UTF-8 bytes sent on the wire.
"""

from __future__ import annotations

import base64
import hashlib
import time
import uuid
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

from sudomimus_token import create_jwt

from .constants import (
    CLIENT_JWT_AUDIENCE,
    CLIENT_JWT_DEFAULT_LIFETIME_SECONDS,
    CLIENT_JWT_MAX_LIFETIME_SECONDS,
)
from .errors import ConnectConfigError

ClientAuthSigner = Callable[[str], str | Awaitable[str]]


@dataclass(frozen=True, slots=True)
class ConnectClientAuthWithKey:
    """Sign client-auth JWTs locally with a PEM-encoded RS256 private key."""

    application_anchor: str
    private_key_pem: str
    lifetime_seconds: int | None = None
    jti_generator: Callable[[], str] | None = None


@dataclass(frozen=True, slots=True)
class ConnectClientAuthWithSigner:
    """Delegate client-auth JWT creation to a bring-your-own signer.

    The signer receives the exact JSON string sent on the wire and MUST
    return a signed JWT whose ``body_sha256`` claim is the standard base64 of
    ``SHA-256(rawBody)``. The signer may be sync or (for the async client)
    return an awaitable.
    """

    application_anchor: str
    signer: ClientAuthSigner


ConnectClientAuth = ConnectClientAuthWithKey | ConnectClientAuthWithSigner


def sha256_base64(raw_body: str) -> str:
    """Standard base64 of ``SHA-256(raw_body)`` over UTF-8 bytes."""
    digest = hashlib.sha256(raw_body.encode("utf-8")).digest()
    return base64.b64encode(digest).decode("ascii")


def build_establish_client_jwt_claims(
    application_anchor: str,
    raw_body: str,
    *,
    lifetime_seconds: int | None = None,
    jti_generator: Callable[[], str] | None = None,
    now: float | None = None,
) -> dict[str, Any]:
    """Build the client-auth JWT claim set without signing it."""
    lifetime = (
        lifetime_seconds
        if lifetime_seconds is not None
        else CLIENT_JWT_DEFAULT_LIFETIME_SECONDS
    )
    if lifetime <= 0 or lifetime > CLIENT_JWT_MAX_LIFETIME_SECONDS:
        raise ConnectConfigError(
            f"clientAuth.lifetime_seconds must be in (0, {CLIENT_JWT_MAX_LIFETIME_SECONDS}]; "
            f"got {lifetime}"
        )

    iat = int(now if now is not None else time.time())
    jti = jti_generator() if jti_generator is not None else str(uuid.uuid4())
    return {
        "iss": application_anchor,
        "aud": CLIENT_JWT_AUDIENCE,
        "iat": iat,
        "exp": iat + lifetime,
        "jti": jti,
        "body_sha256": sha256_base64(raw_body),
    }


def sign_establish_client_jwt(config: ConnectClientAuthWithKey, raw_body: str) -> str:
    """Sign a client-auth JWT (claims in the body, empty header) with RS256."""
    claims = build_establish_client_jwt_claims(
        config.application_anchor,
        raw_body,
        lifetime_seconds=config.lifetime_seconds,
        jti_generator=config.jti_generator,
    )
    return create_jwt({}, claims, config.private_key_pem)
